iree/turbine/kernel/boo/fusion/subgraph.py

```diff
@@ -0,0 +1,181 @@
+    )
+
+    custom_op = get_autograd_function(
+        joint_sg, fake_args_joint, num_fwd_outputs=metadata.num_forward_returns
+    )
+
+    single_node_graph = fused_subgraph(
+        subgraph,
+        custom_op,
+        (n for n in subgraph.graph.nodes if n.op == "placeholder"),
+        num_outputs=metadata.num_forward_returns,
+    )
+    _log_graph_module("Replacement Subgraph", single_node_graph)
+    return single_node_graph
+
+
+def fused_subgraph(
+    root, target: Target, input_nodes: Sequence[Node], num_outputs: int
+) -> GraphModule:
+    g = Graph()
+    new_inputs = []
+    for n in input_nodes:
+        new_node = g.placeholder(n.name)
+        new_node.meta = n.meta
+        new_inputs.append(new_node)
+    fused_node = g.call_function(target, tuple(new_inputs))
+    outputs = []
+    for i in range(num_outputs):
+        outputs.append(g.call_method("__getitem__", args=(fused_node, i)))
+    g.output(result=tuple(outputs))
+    return GraphModule(root, g)
+
+
+def extract_fusion_subgraph_modules(
+    src_gm: GraphModule, fusion_schema: FusionSchema
+) -> Tuple[Sequence[GraphModule], Sequence[Dict[Node, Node]]]:
+    """Traverses src_gm nodes in order. When a node matches a root op in the fusion_schema,
+    A new subgraph is created with the root op in addition to any adjacent nodes matching the schema.
+
+    All subgraphs found this way are returned in a list.
+    The corresponding mappings from src_gm nodes -> subgraph nodes is also returned as a list.
+    """
+    subgraph_projections: List[Dict[Node, Node]] = []
+    subgraph_inclusions: List[Dict[Node, Node]] = []
+    subgraphs: List[GraphModule] = []
+    used_nodes: Set[Node] = set()
+    for root in src_gm.graph.nodes:
+        # if this node is included in any subgraph already, we can't use it.
+        if root in used_nodes:
+            continue
+
+        # TODO: support other situations, e.g. "call_method"
+        if root.op != "call_function":
+            continue
+        # Check if this node is a fusion root op
+        node_spec: OpFusionSpec | None = fusion_schema.get(root.target, None)
+        if node_spec is None:
+            continue
+        node_list = [root]
+
+        # Walk producers from root and include them in the subgraph
+        worklist = [root]
+        while len(worklist) > 0:
+            # We are treating worklist as a FIFO queue (pop from front).
+            # This results in a breadth-first traversal of producers.
+            curr_node = worklist.pop(0)
+            for producer in curr_node.all_input_nodes:
+                if producer.op != "call_function":
+                    continue
+                if producer.target not in node_spec.producers:
+                    continue
+                if producer in used_nodes:
+                    continue
+                # Insert producers at the front, since we want to preserve at least some weak ordering of nodes.
```

Comment:
I'd say we want to have producers in the same order as they appear in the graph, and we are currently reversing this order. It looks possible to construct the following scenario:

```python 
producer1 = call(...)
producer2 = call(producer1). # e.g. compute the mean of producer1
root = call(producer1, producer2) # e.g. normalize dividing by mean
```

where inverting the order of producers looks invalid.

What we need is a topological sort, which can be implemented either as DFS + visited set or as BFS + node degree counters. We could either traverse producers directly in this order, or just sort all nodes once after everything is collected.